demos/DG_advection/DG_advection.py.rst

```diff
@@ -0,0 +1,279 @@
+neglecting the inflow boundary condition.  We also save the initial state so
+that we can check the :math:`L^2`-norm error at the end. ::
+
+  q = Function(V).interpolate(1.0 + bell + cone + slot_cyl)
+  q_init = Function(V).assign(q)
+
+We declare the output filename, and write out the initial condition. ::
+
+  outfile = File("DGadv.pvd")
+  outfile.write(q)
+
+We will run for time :math:`2\pi`, a full rotation.  We take 500 steps, giving
+a timestep close to the CFL limit.  We declare an extra variable ``dtc``; for
+technical reasons, this means that Firedrake does not have to compile new C code
+if the user tries different timesteps.  Finally, we define the inflow boundary
+condition, :math:`q_\mathrm{in}`.  In general, this would be a ``Function``, but
+here we just use a ``Constant`` value. ::
+
+  T = 2*math.pi
+  dt = T/500.0
+  dtc = Constant(dt)
+  q_in = Constant(1.0)
+
+Now we declare our variational forms.  Solving for :math:`\Delta q` at each
+stage, the explicit timestepping scheme means that the left hand side is just a
+mass matrix. ::
+
+  dq_trial = TrialFunction(V)
+  phi = TestFunction(V)
+  a = phi*dq_trial*dx
+
+The right-hand-side is more interesting.  We define ``n`` to be the built-in
+``FacetNormal`` object; a unit normal vector that can be used in integrals over
+exterior and interior facets.  We next define ``un`` to be an object which is
+equal to :math:`\vec{u}\cdot\vec{n}` if this is positive, and zero if this is
+negative.  This will be useful in the upwind terms.  We now define our
+right-hand-side form ``L1`` as :math:`\Delta t` times the sum of four integrals.
+
+The first integral is a straightforward cell integral of
+:math:`q\nabla\cdot(\phi\vec{u})`.  The second integral represents the inflow
+boundary condition.  We only want this to contribute on the inflow part of the
+boundary, where :math:`\vec{u}\cdot\vec{n} < 0` (recall that :math:`\vec{n}` is
+an outward-pointing normal).  Where this is true, the condition gives the
+desired expression :math:`\phi q_\mathrm{in}\vec{u}\cdot\vec{n}`, otherwise the
+condition gives zero.  The third integral operates in a similar way to give
+the outflow boundary condition.  The last integral represents the integral
+:math:`\widetilde{q}(\phi_+ \vec{u} \cdot \vec{n}_+ + \phi_- \vec{u} \cdot \vec{n}_-)`
+over interior facets.  We could again use a conditional in order to represent
+the upwind value :math:`\widetilde{q}` by the correct choice of :math:`q_+` or
+:math:`q_-`, depending on the sign of :math:`\vec{u}\cdot\vec{n_+}`, say.
+Instead, we make use of the quantity ``un``, which is either
+:math:`\vec{u}\cdot\vec{n}` or zero, in order to avoid writing explicit
+conditionals. Although it is not obvious at first sight, the expression given in
+code is equivalent to the desired expression, assuming
+:math:`\vec{n}_- = -\vec{n}_+`. ::
+
+  n = FacetNormal(mesh)
+  un = 0.5*(dot(u, n) + abs(dot(u, n)))
```

Comment:
Move these two lines before this paragraph?
```
This will be useful in the upwind terms. ::

  n = FacetNormal(mesh)
  un = ...

We now define our ...
```